pkg/queryfrontend/queryinstant_codec.go

```diff
@@ -283,22 +302,55 @@ func vectorMerge(resps []*queryrange.PrometheusInstantQueryResponse) *queryrange
 	if len(output) == 0 {
 		return &queryrange.Vector{
 			Samples: make([]*queryrange.Sample, 0),
-		}
+		}, nil
 	}
 
-	keys := make([]string, 0, len(output))
-	for key := range output {
-		keys = append(keys, key)
+	var ss []*queryrange.Sample
+	for _, v := range output {
+		ss = append(ss, v)
 	}
-	sort.Strings(keys)
 
+	sort.Slice(ss, func(i, j int) bool {
+		// Order is determined by the sortFn in the query.
+		if sortAsc {
+			return ss[i].Sample.Value < ss[j].Sample.Value
+		} else if sortDesc {
+			return ss[i].Sample.Value > ss[j].Sample.Value
+		} else {
+			// Fallback on sorting by series
+			m1 := cortexpb.FromLabelAdaptersToLabels(ss[i].Labels).String()
+			m2 := cortexpb.FromLabelAdaptersToLabels(ss[j].Labels).String()
+			return m1 < m2
```

Comment:
This will cause some performance penalties as we need to convert the labels to string again. If no sorting is enabled, can we still keep the previous logic to sort on series keys only?